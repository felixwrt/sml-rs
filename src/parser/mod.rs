//! SML parsers.
//!
//! This module contains parsers for the SML protocol. The `complete` module contains an easy to use allocating
//! parser. The `streaming` module contains a flexible non-allocating parser. See the discussion below for a
//! comparison of the two parsers:
//!
//! # Which parser should I choose?
//!
//! The SML protocol defines two data structures that can hold multiple elements. An SML File can
//! contain multiple SML Messages and the "SML_GetList.Res" message contains a list of values.
//! Because of these two elements, the size of an SML File cannot be known at compile time.
//!
//! The parser in the `complete` module uses dynamic memory allocations `alloc::vec::Vec` for SML
//! Messages and SML Values. This makes the usage straight-forward and if you're using `sml-rs` on
//! a hosted platform, this is most likely the parser you'll want to use.
//!
//! The parser in the `streaming` module works differently and therefore doesn't require dynamic
//! memory allocations. Instead of returning a single data structure representing the whole SML File,
//! this parser produces a stream of events that each have a size known at compile time. Depending on the
//! input, the parser will produce a different number of events, which is how different numbers of SML
//! Messages / Values can be handled. If you're using `sml-rs` on a microcontroller and don't want to use
//! an allocator, this is the parser you'll want to use.
//!
//! # Examples
//!
//! ## Using `complete::parse`
//!
//! ```
//! # #[cfg(feature = "alloc")] {
//! # use sml_rs::parser::complete;
//! let bytes: &[u8] = &[ /*...*/ ];
//!
//! println!("{:#?}", complete::parse(&bytes).expect("error while parsing"));
//! # }
//! ```
//!
//! Output (stripped-down to the relevant parts):
//! ```text
//! File {
//!     messages: [
//!         Message {
//!             message_body: OpenResponse {
//!                 ref_time: SecIndex(23876784),
//!                 ...
//!             },
//!             ...
//!         },
//!         Message {
//!             message_body: GetListResponse {
//!                 val_list: [
//!                     ListEntry { ... },
//!                     ListEntry { ... },
//!                     ListEntry { ... },
//!                 ],
//!             },
//!             ...
//!         },
//!         Message {
//!             message_body: CloseResponse,
//!             ...
//!         },
//!     ],
//! }
//! ```
//!
//! ## Using `streaming::Parser`
//! ```rust
//! # use sml_rs::parser::streaming;
//! let bytes: &[u8] = &[ /*...*/ ];
//!
//! let parser = streaming::Parser::new(bytes);
//! for item in parser {
//!     println!("- {:#?}", item.expect("error while parsing"));
//! }
//! ```
//!
//! Output (stripped-down to the relevant parts):
//! ```text
//! - MessageStart(MessageStart {
//!     message_body: OpenResponse {
//!         ref_time: SecIndex(23876784),
//!         ...
//!     },
//!     ...
//! })
//! - MessageStart(MessageStart {
//!     message_body: GetListResponseStart {
//!         num_values: 3,
//!         ...
//!     },
//!     ...
//! })
//! - ListEntry(ListEntry { ... })
//! - ListEntry(ListEntry { ... })
//! - ListEntry(ListEntry { ... })
//! - GetListResponseEnd(GetListResponseEnd)
//! - MessageStart(MessageStart {
//!     message_body: CloseResponse,
//!     ...
//! })
//! ```
//!
//!

use core::{
    fmt::{self, Debug},
    ops::Deref,
};

use tlf::TypeLengthField;

pub mod common;
#[cfg(feature = "alloc")]
pub mod complete;
mod num;
mod octet_string;
pub mod streaming;
mod tlf;

pub use tlf::TlfParseError;

pub use octet_string::OctetStr;

/// Error type used by the parser
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum ParseError {
    /// There are additional bytes in the input while the parser expects EOF
    LeftoverInput,
    /// The parser expected additional bytes but encountered an EOF
    UnexpectedEOF,
    /// An error occurred while parsing a `TypeLengthField`
    InvalidTlf(TlfParseError),
    /// TLF mismatch while parsing struct / enum
    TlfMismatch(&'static str),
    /// CRC mismatch,
    CrcMismatch,
    /// Expected to find 0x00 as message end marker, got something else
    MsgEndMismatch,
    /// Got a variant id that isn't known. This means it's either invalid or not supported (yet) by the parser
    UnexpectedVariant,
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        <Self as Debug>::fmt(self, f)
    }
}

#[cfg(feature = "std")]
impl std::error::Error for ParseError {}

type ResTy<'i, O> = Result<(&'i [u8], O), ParseError>;
#[allow(dead_code)]
type ResTyComplete<'i, O> = Result<O, ParseError>;

/// SmlParse is the main trait used to parse bytes into SML data structures.
pub(crate) trait SmlParse<'i>
where
    Self: Sized,
{
    /// Tries to parse an instance of `Self` from a byte slice.
    ///
    /// On success, returns the remaining input and the parsed instance of `Self`.
    fn parse(input: &'i [u8]) -> ResTy<Self>;

    /// Tries to parse an instance of `Self` from a byte slice and returns an error if there are leftover bytes.
    ///
    /// On success, returns the parsed instance of `Self`.
    fn parse_complete(input: &'i [u8]) -> ResTyComplete<Self> {
        let (input, x) = Self::parse(input)?;
        if !input.is_empty() {
            return Err(ParseError::LeftoverInput);
        }
        Ok(x)
    }
}

pub(crate) trait SmlParseTlf<'i>
where
    Self: Sized,
{
    fn check_tlf(tlf: &TypeLengthField) -> bool;

    fn parse_with_tlf(input: &'i [u8], tlf: &TypeLengthField) -> ResTy<'i, Self>;
}

impl<'i, T: SmlParseTlf<'i>> SmlParse<'i> for T {
    fn parse(input: &'i [u8]) -> ResTy<Self> {
        let (input, tlf) = TypeLengthField::parse(input)?;
        if !Self::check_tlf(&tlf) {
            return Err(ParseError::TlfMismatch(core::any::type_name::<Self>()));
        }
        Self::parse_with_tlf(input, &tlf)
    }
}

impl<'i, T: SmlParse<'i>> SmlParse<'i> for Option<T> {
    fn parse(input: &'i [u8]) -> ResTy<Self> {
        if let Some(0x01u8) = input.first() {
            Ok((&input[1..], None))
        } else {
            let (input, x) = T::parse(input)?;
            Ok((input, Some(x)))
        }
    }
}

fn take_byte(input: &[u8]) -> ResTy<u8> {
    if input.is_empty() {
        return Err(ParseError::UnexpectedEOF);
    }
    Ok((&input[1..], input[0]))
}

fn take<const N: usize>(input: &[u8]) -> ResTy<&[u8; N]> {
    if input.len() < N {
        return Err(ParseError::UnexpectedEOF);
    }
    Ok((&input[N..], input[..N].try_into().unwrap()))
}

fn take_n(input: &[u8], n: usize) -> ResTy<&[u8]> {
    if input.len() < n {
        return Err(ParseError::UnexpectedEOF);
    }
    Ok((&input[n..], &input[..n]))
}

fn map<O1, O2>(val: ResTy<O1>, mut f: impl FnMut(O1) -> O2) -> ResTy<O2> {
    val.map(|(input, x)| (input, f(x)))
}

struct OctetStrFormatter<'i>(&'i [u8]);

// formats a slice using the compact single-line output even when the parent element should be formatted using "{:#?}"
impl<'i> Debug for OctetStrFormatter<'i> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "{:?}", self.0)
    }
}

struct NumberFormatter<T: Debug, U: Deref<Target = T>>(U);

impl<T: Debug, U: Deref<Target = T>> Debug for NumberFormatter<T, U> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(f, "{:?}{}", self.0.deref(), core::any::type_name::<T>())
    }
}
