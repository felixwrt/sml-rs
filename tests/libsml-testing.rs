use std::iter::FromIterator;
use std::{
    collections::HashSet,
    ffi::{OsStr, OsString},
};

#[test]
fn test_repo_validation() {
    let dir = std::fs::read_dir("./tests/libsml-testing").expect("test folder does not exist");
    let mut bin_filenames = HashSet::new();
    let mut hex_filenames = HashSet::new();
    for entry in dir {
        let entry = entry.unwrap().path();
        match (entry.file_stem(), entry.extension().and_then(OsStr::to_str)) {
            (Some(name), Some("bin")) => {
                bin_filenames.insert(name.to_os_string());
            }
            (Some(name), Some("hex")) => {
                hex_filenames.insert(name.to_os_string());
            }
            _ => {} // ignore other files
        }
    }

    assert_eq!(bin_filenames, hex_filenames);

    if bin_filenames.is_empty() {
        panic!("There are no test files in tests/libsml-testing.\n")
    }

    // check that bin and hex files contain the same content
    for filename in bin_filenames {
        let path = std::path::Path::new("./tests/libsml-testing");

        let bin_path = path.join(OsString::from_iter([
            filename.clone(),
            ".bin".to_string().into(),
        ]));
        let bin_bytes = std::fs::read(bin_path).expect("Couldn't read file");
        let hex_path = path.join(OsString::from_iter([filename, ".hex".to_string().into()]));
        let hex_string = std::fs::read_to_string(hex_path).expect("Couldn't read file");
        let hex_bytes = hex::decode(hex_string.trim()).expect("Couldn't decode hex string");

        assert_eq!(bin_bytes, hex_bytes);
    }
}

#[cfg(feature = "alloc")]
#[test]
fn test_files() {
    use std::fmt::Write;

    insta::glob!("libsml-testing/*.bin", |path| {
        let bytes = std::fs::read(path).unwrap();

        let mut decoder =
            sml_rs::transport::decode_streaming::<sml_rs::util::ArrayBuf<2048>>(bytes);

        let mut s = String::new();
        while let Some(result) = decoder.next() {
            // write!(s, "{:?}\n", result.map(|x| x.len())).unwrap();
            write!(
                s,
                "{:#?}\n",
                result.map(|x| {
                    let res = sml_rs::parser::complete::parse(x);
                    res.expect("Error while parsing:").messages
                })
            )
            .unwrap();
        }
        insta::assert_snapshot!(s);
    });
}

#[cfg(test)]
mod test_attention_response {
    use sml_rs::parser::{
        common::{
            AttentionErrorCode, AttentionNumber, AttentionResponse, CloseResponse, HintNumber,
            OpenResponse, Time, Tree,
        },
        streaming::{self, MessageBody, MessageStart, ParseEvent},
    };

    #[test]
    fn attention_response_order_not_executed() {
        let bytes = vec![
            0x1B, 0x1B, 0x1B, 0x1B, 0x01, 0x01, 0x01, 0x01, 0x76, 0x02, 0x01, 0x62, 0x00, 0x62,
            0x00, 0x72, 0x63, 0x01, 0x01, 0x76, 0x01, 0x07, 0x43, 0x4C, 0x4E, 0x49, 0x44, 0x31,
            0x0A, 0x51, 0x00, 0x00, 0x00, 0x00, 0x66, 0x9F, 0x41, 0xA7, 0x0B, 0x0A, 0x01, 0x4C,
            0x47, 0x5A, 0x00, 0x03, 0xA9, 0xC6, 0x26, 0x72, 0x62, 0x01, 0x65, 0x00, 0x08, 0x5A,
            0xE0, 0x01, 0x63, 0x31, 0x66, 0x00, 0x76, 0x02, 0x02, 0x62, 0x00, 0x62, 0x00, 0x72,
            0x63, 0xFF, 0x01, 0x74, 0x0B, 0x0A, 0x01, 0x4C, 0x47, 0x5A, 0x00, 0x03, 0xA9, 0xC6,
            0x26, 0x07, 0x81, 0x81, 0xC7, 0xC7, 0xFE, 0x0A, 0x01, 0x73, 0x0A, 0x01, 0x00, 0x5E,
            0x31, 0x00, 0x07, 0x00, 0x01, 0x00, 0x01, 0x01, 0x63, 0x5C, 0xF4, 0x00, 0x76, 0x02,
            0x03, 0x62, 0x00, 0x62, 0x00, 0x72, 0x63, 0x02, 0x01, 0x71, 0x01, 0x63, 0xD5, 0x35,
            0x00, 0x00, 0x1B, 0x1B, 0x1B, 0x1B, 0x1A, 0x01, 0xC4, 0x75,
        ];
        let mut decoder =
            sml_rs::transport::decode_streaming::<sml_rs::util::ArrayBuf<2048>>(bytes);
        let expected_response = vec![
            MessageStart {
                transaction_id: &[1],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::OpenResponse(OpenResponse {
                    client_id: Some(&[67, 76, 78, 73, 68, 49]),
                    req_file_id: &[81, 0, 0, 0, 0, 102, 159, 65, 167],
                    codepage: None,
                    server_id: &[10, 1, 76, 71, 90, 0, 3, 169, 198, 38],
                    ref_time: Some(Time::SecIndex(547552)),
                    sml_version: None,
                }),
            },
            MessageStart {
                transaction_id: &[2],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::AttentionResponse(AttentionResponse {
                    server_id: &[10, 1, 76, 71, 90, 0, 3, 169, 198, 38],
                    number: AttentionNumber::AttentionErrorCode(
                        AttentionErrorCode::OrderNotExecuted,
                    ),
                    msg: None,
                    details: Some(Tree {
                        parameter_name: &[0x01, 0x00, 0x5E, 0x31, 0x00, 0x07, 0x00, 0x01, 0x00],
                        parameter_value: None,
                        child_list: None,
                    }),
                }),
            },
            MessageStart {
                transaction_id: &[3],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::CloseResponse(CloseResponse {
                    global_signature: None,
                }),
            },
        ];
        while let Some(result) = decoder.next() {
            let input = result.unwrap();
            let parser = streaming::Parser::new(input);
            for (idx, val) in parser.enumerate().into_iter() {
                if let Ok(ParseEvent::MessageStart(val)) = val {
                    assert_eq!(expected_response[idx], val);
                }
            }
        }
    }

    #[test]
    fn attention_response_positive() {
        let bytes = vec![
            0x1B, 0x1B, 0x1B, 0x1B, 0x01, 0x01, 0x01, 0x01, 0x76, 0x02, 0x01, 0x62, 0x00, 0x62,
            0x00, 0x72, 0x63, 0x01, 0x01, 0x76, 0x01, 0x07, 0x43, 0x4C, 0x4E, 0x49, 0x44, 0x31,
            0x0A, 0x51, 0x00, 0x00, 0x00, 0x00, 0x66, 0x9F, 0x64, 0x3C, 0x0B, 0x0A, 0x01, 0x4C,
            0x47, 0x5A, 0x00, 0x03, 0xA9, 0xC6, 0x26, 0x72, 0x62, 0x01, 0x65, 0x00, 0x08, 0x7D,
            0x56, 0x01, 0x63, 0xC2, 0xF2, 0x00, 0x76, 0x02, 0x02, 0x62, 0x00, 0x62, 0x00, 0x72,
            0x63, 0xFF, 0x01, 0x74, 0x0B, 0x0A, 0x01, 0x4C, 0x47, 0x5A, 0x00, 0x03, 0xA9, 0xC6,
            0x26, 0x07, 0x81, 0x81, 0xC7, 0xC7, 0xFD, 0x00, 0x01, 0x73, 0x0A, 0x01, 0x00, 0x5E,
            0x31, 0x00, 0x07, 0x00, 0x01, 0x00, 0x01, 0x01, 0x63, 0x5B, 0xAF, 0x00, 0x76, 0x02,
            0x03, 0x62, 0x00, 0x62, 0x00, 0x72, 0x63, 0x02, 0x01, 0x71, 0x01, 0x63, 0xD5, 0x35,
            0x00, 0x00, 0x1B, 0x1B, 0x1B, 0x1B, 0x1A, 0x01, 0xAC, 0x0C,
        ];
        let expected_response = vec![
            MessageStart {
                transaction_id: &[1],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::OpenResponse(OpenResponse {
                    client_id: Some(&[67, 76, 78, 73, 68, 49]),
                    req_file_id: &[81, 0, 0, 0, 0, 102, 159, 100, 60],
                    codepage: None,
                    server_id: &[10, 1, 76, 71, 90, 0, 3, 169, 198, 38],
                    ref_time: Some(Time::SecIndex(556374)),
                    sml_version: None,
                }),
            },
            MessageStart {
                transaction_id: &[2],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::AttentionResponse(AttentionResponse {
                    server_id: &[10, 1, 76, 71, 90, 0, 3, 169, 198, 38],
                    number: AttentionNumber::HintNumber(HintNumber::Positive),
                    msg: None,
                    details: Some(Tree {
                        parameter_name: &[0x01, 0x00, 0x5E, 0x31, 0x00, 0x07, 0x00, 0x01, 0x00],
                        parameter_value: None,
                        child_list: None,
                    }),
                }),
            },
            MessageStart {
                transaction_id: &[3],
                group_no: 0,
                abort_on_error: 0,
                message_body: MessageBody::CloseResponse(CloseResponse {
                    global_signature: None,
                }),
            },
        ];
        let mut decoder =
            sml_rs::transport::decode_streaming::<sml_rs::util::ArrayBuf<2048>>(bytes);
        while let Some(result) = decoder.next() {
            let input = result.unwrap();
            let parser = streaming::Parser::new(input);
            for (idx, val) in parser.enumerate().into_iter() {
                if let Ok(ParseEvent::MessageStart(val)) = val {
                    assert_eq!(expected_response[idx], val);
                }
            }
        }
    }
}
