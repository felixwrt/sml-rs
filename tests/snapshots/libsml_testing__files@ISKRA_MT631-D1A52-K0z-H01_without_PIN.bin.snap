---
source: tests/libsml-testing.rs
expression: s
input_file: tests/libsml-testing/ISKRA_MT631-D1A52-K0z-H01_without_PIN.bin
---
Ok(
    [
        Message {
            transaction_id: [9, 157, 243, 236],
            group_no: 0,
            abort_on_error: 0,
            message_body: OpenResponse {
                req_file_id: [3, 52, 166, 164],
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                ref_time: SecIndex(53781751),
                sml_version: 1,
            },
        },
        Message {
            transaction_id: [9, 157, 243, 237],
            group_no: 0,
            abort_on_error: 0,
            message_body: GetListResponse {
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                list_name: [1, 0, 98, 10, 255, 255],
                act_sensor_time: SecIndex(53781751),
                val_list: [
                    ListEntry {
                        obj_name: [1, 0, 96, 50, 1, 1],
                        value: [73, 83, 75],
                    },
                    ListEntry {
                        obj_name: [1, 0, 96, 1, 0, 255],
                        value: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                    },
                    ListEntry {
                        obj_name: [1, 0, 1, 8, 0, 255],
                        status: 1048836u32,
                        unit: 30,
                        scaler: 3,
                        value: 2043u16,
                    },
                    ListEntry {
                        obj_name: [1, 0, 2, 8, 0, 255],
                        unit: 30,
                        scaler: 3,
                        value: 0u8,
                    },
                ],
            },
        },
        Message {
            transaction_id: [9, 157, 243, 238],
            group_no: 0,
            abort_on_error: 0,
            message_body: CloseResponse,
        },
    ],
)
Ok(
    [
        Message {
            transaction_id: [9, 157, 243, 239],
            group_no: 0,
            abort_on_error: 0,
            message_body: OpenResponse {
                req_file_id: [3, 52, 166, 165],
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                ref_time: SecIndex(53781752),
                sml_version: 1,
            },
        },
        Message {
            transaction_id: [9, 157, 243, 240],
            group_no: 0,
            abort_on_error: 0,
            message_body: GetListResponse {
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                list_name: [1, 0, 98, 10, 255, 255],
                act_sensor_time: SecIndex(53781752),
                val_list: [
                    ListEntry {
                        obj_name: [1, 0, 96, 50, 1, 1],
                        value: [73, 83, 75],
                    },
                    ListEntry {
                        obj_name: [1, 0, 96, 1, 0, 255],
                        value: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                    },
                    ListEntry {
                        obj_name: [1, 0, 1, 8, 0, 255],
                        status: 1048836u32,
                        unit: 30,
                        scaler: 3,
                        value: 2043u16,
                    },
                    ListEntry {
                        obj_name: [1, 0, 2, 8, 0, 255],
                        unit: 30,
                        scaler: 3,
                        value: 0u8,
                    },
                ],
            },
        },
        Message {
            transaction_id: [9, 157, 243, 241],
            group_no: 0,
            abort_on_error: 0,
            message_body: CloseResponse,
        },
    ],
)
Ok(
    [
        Message {
            transaction_id: [9, 157, 243, 242],
            group_no: 0,
            abort_on_error: 0,
            message_body: OpenResponse {
                req_file_id: [3, 52, 166, 166],
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                ref_time: SecIndex(53781753),
                sml_version: 1,
            },
        },
        Message {
            transaction_id: [9, 157, 243, 243],
            group_no: 0,
            abort_on_error: 0,
            message_body: GetListResponse {
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                list_name: [1, 0, 98, 10, 255, 255],
                act_sensor_time: SecIndex(53781753),
                val_list: [
                    ListEntry {
                        obj_name: [1, 0, 96, 50, 1, 1],
                        value: [73, 83, 75],
                    },
                    ListEntry {
                        obj_name: [1, 0, 96, 1, 0, 255],
                        value: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                    },
                    ListEntry {
                        obj_name: [1, 0, 1, 8, 0, 255],
                        status: 1048836u32,
                        unit: 30,
                        scaler: 3,
                        value: 2043u16,
                    },
                    ListEntry {
                        obj_name: [1, 0, 2, 8, 0, 255],
                        unit: 30,
                        scaler: 3,
                        value: 0u8,
                    },
                ],
            },
        },
        Message {
            transaction_id: [9, 157, 243, 244],
            group_no: 0,
            abort_on_error: 0,
            message_body: CloseResponse,
        },
    ],
)
Ok(
    [
        Message {
            transaction_id: [9, 157, 243, 245],
            group_no: 0,
            abort_on_error: 0,
            message_body: OpenResponse {
                req_file_id: [3, 52, 166, 167],
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                ref_time: SecIndex(53781754),
                sml_version: 1,
            },
        },
        Message {
            transaction_id: [9, 157, 243, 246],
            group_no: 0,
            abort_on_error: 0,
            message_body: GetListResponse {
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                list_name: [1, 0, 98, 10, 255, 255],
                act_sensor_time: SecIndex(53781754),
                val_list: [
                    ListEntry {
                        obj_name: [1, 0, 96, 50, 1, 1],
                        value: [73, 83, 75],
                    },
                    ListEntry {
                        obj_name: [1, 0, 96, 1, 0, 255],
                        value: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                    },
                    ListEntry {
                        obj_name: [1, 0, 1, 8, 0, 255],
                        status: 1048836u32,
                        unit: 30,
                        scaler: 3,
                        value: 2043u16,
                    },
                    ListEntry {
                        obj_name: [1, 0, 2, 8, 0, 255],
                        unit: 30,
                        scaler: 3,
                        value: 0u8,
                    },
                ],
            },
        },
        Message {
            transaction_id: [9, 157, 243, 247],
            group_no: 0,
            abort_on_error: 0,
            message_body: CloseResponse,
        },
    ],
)
Ok(
    [
        Message {
            transaction_id: [9, 157, 243, 248],
            group_no: 0,
            abort_on_error: 0,
            message_body: OpenResponse {
                req_file_id: [3, 52, 166, 168],
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                ref_time: SecIndex(53781755),
                sml_version: 1,
            },
        },
        Message {
            transaction_id: [9, 157, 243, 249],
            group_no: 0,
            abort_on_error: 0,
            message_body: GetListResponse {
                server_id: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                list_name: [1, 0, 98, 10, 255, 255],
                act_sensor_time: SecIndex(53781755),
                val_list: [
                    ListEntry {
                        obj_name: [1, 0, 96, 50, 1, 1],
                        value: [73, 83, 75],
                    },
                    ListEntry {
                        obj_name: [1, 0, 96, 1, 0, 255],
                        value: [10, 1, 73, 83, 75, 0, 4, 122, 85, 68],
                    },
                    ListEntry {
                        obj_name: [1, 0, 1, 8, 0, 255],
                        status: 1048836u32,
                        unit: 30,
                        scaler: 3,
                        value: 2043u16,
                    },
                    ListEntry {
                        obj_name: [1, 0, 2, 8, 0, 255],
                        unit: 30,
                        scaler: 3,
                        value: 0u8,
                    },
                ],
            },
        },
        Message {
            transaction_id: [9, 157, 243, 250],
            group_no: 0,
            abort_on_error: 0,
            message_body: CloseResponse,
        },
    ],
)

